Documentation for developers to understand continuous build integration (#990)
OpenLiberty/liberty-tools-intellij: docs/LSP4IJ-Continuous-Integration.md

**call-build-workflow-for-each-merge-commit-sha:**

- Triggers a build workflow (.github/workflows/build.yaml) for each merge commit SHA from a valid PR that does not have merge conflicts.

- Depends on fetch_all_pull_request_shas. Runs only if is_empty is false.

- Sets the value of useLocalPlugin to true so that the build workflow uses a local build of LSP4IJ and not LSP4IJ from the marketplace.

**call-build-workflow-for-lsp4ij-main-branch:**

- Runs the build workflow for the LSP4IJ main branch.

- refLsp4ij: Set to “main” to run the build using the LSP4IJ main branch.

**call-build-workflow-slack-notification:**

- Runs on “ubuntu-latest”.

- Determines the final status based on the results of the build jobs.

- Sends a Slack notification with the workflow status and build results.

- Depends on both call-build-workflow-for-each-merge-commit-sha and call-build-workflow-for-lsp4ij-main-branch.

- Always runs regardless of previous job results.

## Results from the Cron Job

**Example of a Cron Job Output** - [Here](https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/11011078568The) 

Each build runs on Linux, Mac, and Windows.

> There are 4 jobs in the build. The first job, **PR details**, outputs a list of PRs in the LSP4IJ repository and displays warnings for any PRs that are either drafts or have merge conflicts.

Reviewer: ```suggestion
There are four jobs in the build. The first job, **PR details**, outputs a list of PRs in the LSP4IJ repository and displays warnings for any PRs that are either drafts or have merge conflicts.
```

<img alt="Pr-details" height="350" src="images/Pr-details.png" width="400" style="display: block; margin: 0 auto;"/>

> The next job **Run Lsp4ij Main** is the build which runs against the LSP4IJ main branch by checking out the LSP4IJ main branch. The job **Run PR** runs by checking out the merge commit SHAs of each PR listed in the job **PR details**. We can see the LSP4IJ PR number against which it is run in the bracket, followed by the merge commit SHA. After the above three jobs are completed, **Run Slack Notification** job runs and we get the Build results as slack notification.

Reviewer: ```suggestion
The next job, **Run Lsp4ij Main**, is the build which runs against the LSP4IJ main branch by checking out the LSP4IJ main branch. The job **Run PR** runs by checking out the merge commit SHAs of each PR listed in the job **PR details**. The LSP4IJ PR number against which the job is run is displayed in the bracket, followed by the merge commit SHA. After the above three jobs are completed, the **Run Slack Notification** job runs and the Build results are sent as slack notification.
```

![Result](images/result-cron-job.png)

> In the Annotations section, we get the details of the failed tests and also warnings of PRs with merge conflict.

Reviewer: ```suggestion
The Annotations section contains errors that show the details of the failed builds and also warnings that show PRs with merge conflicts.
```

![Warnings](images/Warnings-PR.png)

![Build-failed](images/Build-failed.png)

# Testing against a Specific Version of LSP4IJ

In order to run the LTI build against a particular version of LSP4IJ (stable or nightly), the version can be set in the build.gradle file.

![Version](images/version.png)

To run with a nightly version (For example: 0.4.0-20240813-173701 ), change the line `def lsp4ijVersion = '0.3.0'` to `def lsp4ijVersion = '0.4.0-20240813-173701@nightly'`. Make sure to add **@nightly** to the version number while testing with nightly versions. For stable versions, just specify the version number.

> LSP4IJ nightly builds can be downloaded from the [JetBrains Marketplace](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly)

Reviewer: ```suggestion
LSP4IJ nightly builds can be downloaded from the [JetBrains Marketplace](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly).
```
